### PR TITLE
Added Compose UI Test for DownloadBookItem, OnlineBookItem

### LINKED ISSUE #4855: Add test cases for `OnlineBookItem`, and `DownloadBookItem`
These are pure compose UI, and we should test this UI with `Robolectric`.

### PR DESCRIPTION
**Pull request by atharvyadav22**
Fixes: #4855

### CODE REVIEW

**Comment by atharvyadav22** [comment]
Will update on this soon only 1-2 tests are left and few errors otherwise i have added more semantics to properly handle

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz 
As these are already marked as flaky so would work on that fixes too completely.
But have a look at the PR

**Comment by MohitMaliFtechiz** [comment]
> As these are already marked as flaky so would work on that fixes too completely.

@atharvyadav22 Exactly which one is marked as flaky?

**Comment by atharvyadav22** [comment]
> > As these are already marked as flaky so would work on that fixes too completely.
> 
> @atharvyadav22 Exactly which one is marked as flaky?

The Failing CI like CategoryViewModelTest, etc 
I'll work on that too as there as several `flakyTest{} ` in codebase

**Comment by atharvyadav22** [comment]
After This Merges #4787 in the main this issue would be solve as `TestApplication` is being moved to `sharedFunctions` so to avoid regression will rebase it after the merge.

**Comment by gouri-panda** [comment]
@atharvyadav22 What's the status on this? Should we move?

**Comment by atharvyadav22** [comment]
> @atharvyadav22 What's the status on this? Should we move?

Yes, The CI is failing for different reasons not releated to this PR wouldfix all that flaky tests soon.

**Comment by MohitMaliFtechiz** [comment]
@atharvyadav22
* Now the flaky test cases for `CategoryViewModel` and `LanguageViewModel` are fixed. 
* We have removed the `ZimManageViewModelTest` completely from the project. Since it has been migrated to `LocalLibraryViewModel` and `OnlineLibraryViewModel`.

So now this PR is ready for review?

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz @gouri-panda 
Please have a look!

File: app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+      Byte(downloaded.toString()).humanReadable + "/" + Byte(total.toString()).humanReadable
+    composeTestRule
+      .onNode(hasText(expectedText))
+      .assertIsDisplayed()
+  }
+
+  @Test
+  fun downloadBookItem_whenDownloadPaused_pauseResumeButtonIsDisplayed() {
+    setContent(item = mockLibraryDownloadItem(downloadState = DownloadState.Paused))
+    composeTestRule
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt:287]: @atharvyadav22 We are here only checking that the icon is displayed or not. But we should also assert which icon is showing, e.g., pause or resume, to ensure the correct UI is showing.

File: app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+      .onNode(hasTestTag(DOWNLOADING_PAUSE_BUTTON_TESTING_TAG))
+      .assertIsDisplayed()
+  }
+
+  @Test
+  fun downloadBookItem_whenDownloadRunning_pauseResumeButtonIsDisplayed() {
+    setContent(item = mockLibraryDownloadItem(downloadState = DownloadState.Running))
+    composeTestRule
+      .onNode(hasTestTag(DOWNLOADING_PAUSE_BUTTON_TESTING_TAG))
+      .assertIsDisplayed()
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt:297]: Same for this test.

File: app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+  @Test
+  fun downloadBookItem_whenFailedWithConnectionTimedOut_onStopClickCalledByLaunchedEffect() {
+    setContent(
+      item = mockLibraryDownloadItem(
+        currentDownloadState = Status.FAILED,
+        downloadState = DownloadState.Failed(Error.CONNECTION_TIMED_OUT, null),
+        downloadError = Error.CONNECTION_TIMED_OUT
+      )
+    )
+    verify(exactly = 1) { mockOnStopClick.invoke(any()) }
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt:355]: Good test 👏 . But we should also test that it will only call once, even when recomposition happens.

```kotlin
@Test
  fun downloadBookItem_whenRecomposedWithSameKeys_onStopClickCalledOnlyOnce() {
    val item = mockLibraryDownloadItem(
      id = 1L,
      currentDownloadState = Status.FAILED,
      downloadState = DownloadState.Failed(Error.UNKNOWN_IO_ERROR, null),
      downloadError = Error.UNKNOWN_IO_ERROR
    )

    lateinit var triggerRecompose: MutableState<Int>

    composeTestRule.setContent {
      triggerRecompose = remember { mutableStateOf(0) }
      triggerRecompose.value

      DownloadBookItem(
        index = testIndex,
        item = item,
        onPauseResumeClick = mockOnPauseResumeClick,
        onStopClick = mockOnStopClick
      )
    }

    composeTestRule.waitForIdle()

    verify(exactly = 1) { mockOnStopClick.invoke(item) }

    composeTestRule.runOnIdle {
      triggerRecompose.value++
    }

    composeTestRule.waitForIdle()

    verify(exactly = 1) { mockOnStopClick.invoke(item) }
  }
```

File: app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    setContent(
+      item = mockLibraryDownloadItem(
+        currentDownloadState = Status.FAILED,
+        downloadState = DownloadState.Failed(Error.NONE, null),
+        downloadError = Error.NONE
+      )
+    )
+    verify(exactly = 0) { mockOnStopClick.invoke(any()) }
+  }
+
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt:417]: We should test the state update as well.

```kotlin
@Test
  fun downloadBookItem_whenDownloadStateChanges_stateTextUpdates() {
    lateinit var itemState: MutableState<LibraryDownloadItem>

    val runningItem = mockLibraryDownloadItem(
      downloadState = DownloadState.Running,
      eta = Seconds(90L)
    )

    composeTestRule.setContent {
      itemState = remember { mutableStateOf(runningItem) }

      DownloadBookItem(
        index = testIndex,
        item = itemState.value,
        onPauseResumeClick = mockOnPauseResumeClick,
        onStopClick = mockOnStopClick
      )
    }

    composeTestRule.waitForIdle()

    composeTestRule
      .onNode(
        hasContentDescription(
          runningItem.readableEta.toString() + runningItem.hashCode()
        )
      )
      .assertIsDisplayed()

    val pausedItem = runningItem.copy(
      downloadState = DownloadState.Paused
    )

    composeTestRule.runOnIdle {
      itemState.value = pausedItem
    }

    composeTestRule.waitForIdle()
    val expectedPausedDescription =
      DownloadState.Paused.toReadableState(context).toString() +
        pausedItem.hashCode()

    composeTestRule
      .onNode(hasContentDescription(expectedPausedDescription))
      .assertIsDisplayed()
  }
```

File: app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+  fun downloadBookItem_bookDescriptionIsDisplayed() {
+    setContent(item = mockLibraryDownloadItem(description = "The free encyclopedia"))
+    composeTestRule
+      .onNode(hasContentDescription("The free encyclopedia$testIndex"))
+      .assertIsDisplayed()
+  }
+
+  @Test
+  fun downloadBookItem_whenDescriptionNull_orEmptyBranchRendersWithoutCrash() {
+    setContent(item = mockLibraryDownloadItem(description = null))
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt:155]: This only verifies if the description is null. We should test with an empty description as well.

```kotlin
  @Test
  fun downloadBookItem_whenDescriptionIsEmpty_emptyDescriptionIsDisplayed() {
    setContent(item = mockLibraryDownloadItem(description = ""))

    composeTestRule
      .onNode(hasContentDescription(testIndex.toString()))
      .assertExists()
  }
```

File: app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+  @Test
+  fun downloadBookItem_bookIconIsDisplayed() {
+    val favicon = "https://kiwix.org/favicon.png"
+    setContent(item = mockLibraryDownloadItem(favIconUrl = favicon))
+    val expectedDesc =
+      context.getString(R.string.fav_icon) + favicon.hashCode()
+    composeTestRule
+      .onNode(hasContentDescription(expectedDesc))
+      .assertIsDisplayed()
+  }
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/DownloadBookItemUITest.kt:475]: @atharvyadav22 There are 2 methods in `DownloadBookItem` that should be tested as well, with all edge cases.

```kotlin
private fun getDownloadStateText(item: LibraryDownloadItem, context: Context): String {
  val currentDownloadState = item.downloadState
  return if (currentDownloadState == DownloadState.Running) {
    item.readableEta.toString()
  } else {
    currentDownloadState.toReadableState(context).toString()
  }
}

private fun getDownloadedSizeText(downloadedBytes: Long, totalBytes: Long): String {
  if (totalBytes <= 0 || downloadedBytes <= 0) return ""
  val downloadedText = Byte(downloadedBytes.toString()).humanReadable
  val totalText = Byte(totalBytes.toString()).humanReadable
  return "$downloadedText/$totalText"
}
```

File: app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineBookItemUITest.kt [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+        index = ZERO,
+        item = item,
+        bookUtils = mockBookUtils,
+        availableSpaceCalculator = mockAvailableSpaceCalculator,
+        onBookItemClick = onBookItemClick
+      )
+    }
+    composeTestRule.waitForIdle()
+  }
+
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineBookItemUITest.kt:141]: We should verify that availableSpaceForBook is called when rendering the books.

```kotlin
  @Test
  fun onlineBookItem_checksAvailableSpaceWhenRendered() {
    val book = mockLibkiwixBook()
  
    setContent(item = mockBookItem(book = book))

    coVerify(exactly = 1) {
      mockAvailableSpaceCalculator.hasAvailableSpaceForBook(book)
    }
  }
```

File: app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineBookItemUITest.kt [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+  @Test
+  fun onlineBookItem_whenDetectingFileSystem_overlayExistsInTree() {
+    coEvery {
+      mockAvailableSpaceCalculator.hasAvailableSpaceForBook(any())
+    } returns false
+    setContent(item = mockBookItem(fileSystemState = DetectingFileSystem))
+    composeTestRule
+      .onNodeWithTag(ONLINE_BOOK_OVERLAY_TESTING_TAG)
+      .assertExists()
+  }
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/nav/destination/library/online/OnlineBookItemUITest.kt:392]: These are good tests, but we should also test when storage changes.

```kotlin
 @Test
  fun onlineBookItem_overlayVisibilityUpdatesWhenAvailableSpaceChanges() {
    val book = mockLibkiwixBook()

    var item by mutableStateOf(
      mockBookItem(book = book)
    )

    coEvery {
      mockAvailableSpaceCalculator.hasAvailableSpaceForBook(book)
    } returns false

    composeTestRule.setContent {
      OnlineBookItem(
        index = ZERO,
        item = item,
        bookUtils = mockBookUtils,
        availableSpaceCalculator = mockAvailableSpaceCalculator,
        onBookItemClick = mockOnBookItemClick
      )
    }

    composeTestRule.waitForIdle()

    composeTestRule
      .onNodeWithTag(ONLINE_BOOK_OVERLAY_TESTING_TAG)
      .assertExists()

    coEvery {
      mockAvailableSpaceCalculator.hasAvailableSpaceForBook(book)
    } returns true

    composeTestRule.runOnUiThread {
      item = item.copy(
        fileSystemState = DetectingFileSystem
      )
    }

    composeTestRule.waitForIdle()

    composeTestRule
      .onNodeWithTag(ONLINE_BOOK_OVERLAY_TESTING_TAG)
      .assertDoesNotExist()
  }
```

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]
@atharvyadav22 Thank you for adding the test cases. Please see the review changes.

**Review by MohitMaliFtechiz — APPROVED** [review]
@atharvyadav22 Thank you 🎉 . LGTM.
